rfcs/records.md

```diff
@@ -0,0 +1,201 @@
+metatables that define methods, sometimes they overload operators, sometimes they are just used to store data.
+
+Tables aren't perfect for objects; this proposal is motivated by the following deficiencies in using tables-as-objects (in rough priority order):
+
+1. Memory overhead. An object with 6 fields takes `8+56+8+8*32=328` bytes of storage.
+This is because tables have a large header, the hash portion is rounded to a power of two, and every entry has key and value despite the fixed object structure.
+2. Type system compatibility. While Luau type checker can type tables, including ones with rigid structure, it falls short in OOP scenarios because it's very difficult
+to associate methods with table structure in idiomatic OOP in Luau with tables. Today the situation is especially dire because each method gets its own inferred self type
+(something that is likely to change), and it's impossible to specify a table-with-metatable type via obvious type syntax.
+3. Runtime overhead. While we heavily optimize table reads and writes in object-like scenario, there's still a non-zero cost that is paid for establishing the mapping
+from the index literal to the hash field. Method calls are similarly optimized but the optimization is restricted by the generality of table structure. Finally, large
+object size results in higher cache pressure which affects performance.
+4. Strictness of access. Today at runtime, reading an unknown key from the table returns `nil` and writing a new key just works. This creates significant amount of
+complexity in the type checker, as it has to differentiate between tables that are open for extension and closed via a set of heuristics, and results in easy to make mistakes in untyped code.
+
+Today, writing idiomatic objects in Luau is relatively straightforward but making idiomatic OOP type safe or maximally efficient is next to impossible.
+
+## Design
+
+This proposal suggests solving these problems with a new data type, called record. Record is a type that is semantically a dictionary with a fixed set of keys; much like
+a table, values associated with keys can be read or written to. Much like a table, it has a metatable that can be used to customize behavior of the object by providing
+extra operators (for arithmetics, stringification, etc.), as well as specifying methods.
+
+The difference between records and tables is that when creating a record, the VM allocates space for all values of all keys without having to duplicate the key data;
+since the structure of a record is immutable even if the contents isn't, this allows a much more efficient representation. A 6-field record will take estimated `8+32+6*16=136`
+bytes without implementation heroics, which is more than twice as efficient as table storage. For applications that use many objects this has potential to halve
+the memory footprint.
```

Comment:
Yeah that's correct. There's avenues for optimization here to reorder the checks dynamically via bytecode patching.

Reply:
Err actually sorry, that's not quite correct. But the details here are kinda beyond the scope. We don't actually need to compare the shape pointers when fetching fields, we only need to compare that the name of the record field with the predicted index matches our expectations. The predicted index can be updated if the speculation fails. The name comparison is just a pointer comparison. This is similar to what we do for tables but can be faster because of a few tiny details that would require a very wide margin to explain.

Reply:
Yes and having immutable shapes makes that much easier.

Reply:
Oh our comments are overtaking each other.

Doesn't doing the name comparison require keeping the name in the record? That's an extra word of data per field isn't it? (Unless you're planning on doing some clever bit-twiddling to pack it in with the tag.) Keeping the name in the shape has the advantage that (hopefully) the same shape is shared among a lot of tables.

You also need to do a bounds check if you're going to access the property offset directly.

Reply:
> Doesn't doing the name comparison require keeping the name in the record?
No - the name is kept in the shape. And yeah you need a bounds check.